Feature: shared ownership

cap10morgan opened the pull request:
Implements #31.

Adds new `team (add | list | remove)` and `whoami` commands. Asks for username on init and stores it in git config. A global default username can be configured with `git config --global dgit.username [username]`, but it also looks for any configured GitHub username and suggests that for your dgit username. It then suggests `username/basedir` for the full repo name.

It can also migrate the old `dgit.default` keyring entries to the new `dgit.[username]` entries.

And a few other small fixes & improvements (which I will pull into their own PR if controversial).

Tests shmests... ;)

cap10morgan committed:
DRY up Makefile a bit

cap10morgan committed:
Add basic username init, storage, & key scoping

I had to switch to a fork of go-git for this to pull in a pending PR (on the old src-d GitHub repo) to add global git config support.

cap10morgan committed:
Add team management commands

add, list, and remove

cap10morgan committed:
Move more boilerplate into shared cmd.openRepo

cap10morgan committed:
Add a whoami subcommand to get your username

These come from the git config so they can vary per repo

cap10morgan committed:
Also add collaborators' keyaddrs to repoTree owners

...so they can actually push to the repo

cap10morgan committed:
Fix imports in cmd/team.go

cap10morgan committed:
Migrate old dgit.default key to new dgit.username key

cap10morgan committed:
Rename dgit.New to NewRepo & use it

cap10morgan committed:
Use configured github username as dgit username default

cap10morgan committed:
Default repo name to username/basedir

cap10morgan committed:
Fix username config storage

cap10morgan committed:
Pass new owner key to repotree.Create

cap10morgan committed:
Tiny changes

cap10morgan committed:
go fmt ./...

cap10morgan committed:
Add basic team cmd docs to README

cap10morgan committed:
Merge branch 'master' into feature/shared-ownership

brandonwestcott reviewed (approved):
A few questions / comments, but lgtm 👍 

It seems the key migration is handled automatically? Are there any other effects or manual steps for existing users?

brandonwestcott commented:
Also had a thought on this over the weekend, but needs discussion and certainly can be a future improvement:

What if we stored the team membership in files in the repo, then on push to the HEAD branch (still needs implemented) we updated the tupelo authentications array? Then any future branch-limiting conditions, such as master can only be updated after # thumbs on a PR, will benefit from that protection.

cap10morgan committed:
Use new Merged config feature I wrote for go-git

Phew... that ended being a bit more work than I'd hoped. But it works now, yay!

cap10morgan commented:
> A few questions / comments, but lgtm 👍
> 
> It seems the key migration is handled automatically? Are there any other effects or manual steps for existing users?

Yes it should be. The other effect is that if the user sets a dgit username that is different than what's in their `~/.gitconfig` file, it will be written to the local repo's `./.git/config` file.

cap10morgan commented on initializer/initializer.go:
```diff
@@ -103,41 +174,62 @@ func (i *Initializer) getAuth() (transport.AuthMethod, error) {
 	if isNew {
+		ctx := context.Background()
+
+		opts := &usertree.Options{
```
Yeah probably. I can try it out.

cap10morgan replied:
Done in e5d7671.

cap10morgan commented on initializer/initializer.go:
```diff
@@ -84,14 +99,70 @@ func (i *Initializer) Init(ctx context.Context, args []string) error {
+
+	// try looking up the GitHub username for the default
+	if username == "" {
+		if ghConfig := repoConfig.Raw.Section("github"); ghConfig != nil {
```
It is! Various git tools I use look for it (hub, Tower, etc.).

cap10morgan commented on transport/dgit/client.go:
```diff
@@ -50,44 +50,133 @@ func Default() (*Client, error) {
 	return server.NewServer(loader).NewReceivePackSession(ep, auth)
 }
+
+func (c *Client) AddRepoCollaborator(ctx context.Context, repo *Repo, collaborators []string) error {
```
I can try this out too and see how it compares.

cap10morgan replied:
Looked into this quickly, and I think there definitely could be something to it. The conceptual boundaries of the new `Repo` stuff is still in flux for sure. But I think I'd want to save it for a potential future refactor at this point.

cap10morgan committed:
Refactor (user|repo)tree.Options to alias namedtree.Options

cap10morgan committed:
Allow env username DGIT_USERNAME

...and fix some test issues

cap10morgan committed:
Document username config in README

brandonwestcott commented on transport/dgit/client.go:
```diff
@@ -50,44 +49,131 @@ func Default() (*Client, error) {
-	dir := path.Join(basePath, protocol)
-	c.tupelo, c.nodestore, err = clientbuilder.Build(ctx, dir)
+	dir := path.Join(basePath, constants.Protocol)
+	c.Tupelo, c.Nodestore, err = clientbuilder.Build(ctx, dir)
```
@cap10morgan just to note related to this, I added `Client.DagStore()` method in https://github.com/quorumcontrol/tupelo-go-sdk/pull/212/files#diff-7f6280b5f587f41a183254e1b182f3c9R100 so we don't have to keep passing around both vars everywhere

Commit:
Fix tests with new user chaintrees

cap10morgan committed:
Use better method for setting username in config
